device/components/RTA_OnDeviceComponentTask.brs

```diff
@@ -79,88 +143,88 @@ sub runTaskThread()
-			else if messageType = "roUrlEvent" then
-				id = message.getSourceIdentity().toStr()
-				http = m.activeTransfers[id]
-				m.activeTransfers.delete(id)
-				logDebug("Sent callback to: " + http.getUrl() + " and received response code: " + message.getResponseCode().toStr())
 			else
 				logWarn(messageType + " type not handled")
 			end if
 		end if
 	end while
 end sub
 
-sub verifyAndHandleRequest(receivedString as String, socket as Object)
-	request = parseJson(receivedString)
-	if NOT isAA(request) then
-		logError("Received message did not contain valid request " + receivedString)
-		return
-	end if
-
-	setLogLevel(getStringAtKeyPath(request, "settings.logLevel"))
-
-	if (NOT isInteger(request.callbackPort)) then
-		logError("Received message did not have callbackPort " + receivedString)
-		return
-	end if
+function receiveDataForRequest(request as Object, bufferLength as Integer) as Boolean
+	socket = request.socket
+	' Check if we are going to receive binary or string data
+	if request.stringPayload.len() = request.stringLength then
+		if bufferLength > 0 then
+			ba = createObject("roByteArray")
+			ba[bufferLength - 1] = 0
+			socket.receive(ba, 0, bufferLength)
+			request.binaryPayload.append(ba)
+		end if
 
-	requestId = request.id
-	if (NOT isNonEmptyString(requestId)) then
-		logError("Received message did not have id " + receivedString)
-		return
+		if request.binaryLength = request.binaryPayload.count() then
+			return true
+		end if
+		return false
+	else
+		' Figure out amount to pull from the buffer for string
+		if bufferLength > request.stringLength then
+			receiveLength = request.stringLength
+		else
+			receiveLength = bufferLength
+		end if
+		request["stringPayload"] += socket.receiveStr(receiveLength)
+		bufferLength -= receiveLength
+		return receiveDataForRequest(request, bufferLength)
 	end if
+end function
 
-	receiveAddress = socket.getReceivedFromAddress()
-	request["callbackHost"] = receiveAddress.getHostName()
 
-	socket.setSendToAddress(receiveAddress)
-	bytesSent = socket.sendStr(formatJson({
-		"id": requestId
-	}))
-	if bytesSent = -1 then
-		logError("Could not send ack back")
+sub verifyAndHandleRequest(request)
+	json = parseJson(request.stringPayload)
+	if NOT isAA(json) then
+		logError("Received message did not contain valid request " + request.stringPayload)
 		return
 	end if
+	request.json = json
 
-	componentVersion = getVersion()
-	requestVersion = getStringAtKeyPath(request, "version")
-
-	if requestVersion <> componentVersion then
-		sendBackError(request, "Request version " + requestVersion + " did not match component version " + componentVersion)
+	requestId = json.id
+	if NOT isNonEmptyString(requestId) then
+		logError("Received message did not have id " + request.stringPayload)
 		return
 	end if
 
-	if NOT isAA(request.args) then
-		sendBackError(request, "No args supplied for request type '" + requestType + "'")
+	' TODO look into changing how we do this to avoid overhead associated
```

Comment:
Will this TODO be resolved as part of this PR?

Reply:
No. If and when I get around to it it will be separate. Just adding a TODO as a reminder.